File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int leastPriceSoFar = Integer.MAX_VALUE;
        int profit = 0;
        int profitIfSoldToday = 0;
        
        for(int i = 0; i < prices.length; i++) {
            if(prices[i] < leastPriceSoFar) {
                leastPriceSoFar = prices[i];
            }
            profitIfSoldToday = prices[i] - leastPriceSoFar;
            if(profit < profitIfSoldToday) {
                profit = profitIfSoldToday;
            }
        }
        return profit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        
        int maxProfit = 0;
        
        int buy = 0, sell = 1;
        
        while (sell < prices.size()) {
            if (prices[sell] > prices[buy]) {
                int profit = prices[sell] - prices[buy];
                maxProfit = max(profit, maxProfit);          
            } else {
                buy = sell;
            }
            sell++;
        }
        return maxProfit;
    }
};